### PR TITLE
HW2 have been created

### PR DESCRIPTION
**Pull request by DimaDziublenko**

### CODE REVIEW

File: SQLservise/src/main/java/com/itea/tasks/CRUDServices/CompanyService.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package com.itea.tasks.CRUDServices;
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/CRUDServices/CompanyService.java]: rename CRUDServices to services

File: SQLservise/src/main/java/com/itea/tasks/entities/Company.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.JoinColumn;
+import javax.persistence.JoinTable;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "companies")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Company.java]: company

File: SQLservise/src/main/java/com/itea/tasks/entities/Company.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+@NoArgsConstructor
+@Setter
+@Getter
+public class Company extends MappedSuperClass {
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Company.java]: MappedSuperClass -> BaseEntity

File: SQLservise/src/main/java/com/itea/tasks/entities/Company.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Column(name = "title")
+    private String title;
+
+    @javax.persistence.Column(name = "industry")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Company.java]: @javax.persistence.Column -> @Column

File: SQLservise/src/main/java/com/itea/tasks/entities/Developer.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    private String name;
+
+    @Column(name = "age")
+    private Integer age;
+
+    @Column(name = "gender")
+    private String gender;
+
+    @Column(name = "salary")
+    private Float salary;
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Developer.java]: Use Double

File: SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQuery.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    private static Query query;
+
+    public Project selectTheMostExpensiveProjectBasedOnDevelopersSalary() {
+        query = em.createQuery("SELECT p.id, p.title, p.industry, p.deadline\n" +
+                "FROM projects AS p\n" +
+                "INNER JOIN developer_project AS dp ON p.id = dp.project_id\n" +
+                "INNER JOIN developers AS d ON dp.developer_id = d.id\n" +
+                "GROUP BY p.id, dp.developer_id, dp.project_id, d.id\n" +
+                "ORDER BY SUM(d.salary) DESC");
+        return (Project) query.getSingleResult();
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQuery.java]: use getResultList() everywhere

File: SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQueryTest.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package com.itea.tasks.entities;
+
+import static org.junit.Assert.assertNotNull;
+
+public class EntitiesQueryTest {
+
+    private static EntitiesQuery entitiesQuery = new EntitiesQuery();
+
+    public void selectTheMostExpensiveProjectBasedOnDevelopersSalaryShouldWorkCorrectly() {
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQueryTest.java]: @Test ?

File: SQLservise/src/main/java/com/itea/tasks/entities/Project.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+public class Project extends MappedSuperClass {
+
+    @Column(name = "title")
+    private String title;
+
+    @Column(name = "industry")
+    private String industry;
+
+    @Column(name = "cost")
+    private Float cost;
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Project.java]: Double

File: SQLservise/src/main/java/com/itea/tasks/entities/Project.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    @ManyToMany(fetch = FetchType.EAGER, mappedBy = "projects")
+    private Set<Developer> developers;
+
+    @ManyToMany(fetch = FetchType.EAGER, mappedBy = "projects")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Project.java]: For what do you use EAGER?

File: SQLservise/src/main/java/com/itea/tasks/injectRandomValue/InjectRandomValueClass.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+                if (min > 0)
+                    value = (int) (min + random() * (max - min));
+                else
+                    value = (int) (min + random() * (abs(min) + max));
+                if (value != 0 && !field.isAnnotationPresent(InjectRandomValue.class)) {
+                    if (field.getType() == String.class)
+                        field.set(object.getClass(), "unknown");
+                    if (field.getType() == char.class)
+                        field.setChar(object.getClass(), (char) value);
+                    if (field.getType() == boolean.class)
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/injectRandomValue/InjectRandomValueClass.java]: add branches everywhere
if() {
...
}

File: SQLservise/src/main/resources/init.sql [hunk 11]
```diff
@@ -0,0 +1,64 @@
+CREATE TABLE developers (
+    id     bigint,
+    name   varchar(30),
```
Comment (Aleksandr04) [inline SQLservise/src/main/resources/init.sql]: for varchar use the next numbers: 32, 64, 124

File: SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQueryTest.java (hunk 7 above)
Comment (DimaDziublenko) [reply SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQueryTest.java]: forgot, sorry

File: SQLservise/src/main/java/com/itea/tasks/entities/Project.java (hunk 9 above)
Comment (DimaDziublenko) [reply SQLservise/src/main/java/com/itea/tasks/entities/Project.java]: I was watching the exaple and i saw the man using eager, is it ok or not?

File: SQLservise/src/main/java/com/itea/tasks/entities/BaseEntity.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+import lombok.Getter;
+
+import javax.persistence.Id;
+import javax.persistence.MappedSuperclass;
+
+@MappedSuperclass
+@Getter
+public class BaseEntity {
+
+    @Id
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/BaseEntity.java]: Add @GenaratedValue

File: SQLservise/src/main/java/com/itea/tasks/entities/Company.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "company")
+@NoArgsConstructor
+@Setter
+@Getter
+public class Company extends BaseEntity {
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Company.java]: CompanyEntity

File: SQLservise/src/main/java/com/itea/tasks/entities/Customer.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "customers")
+@NoArgsConstructor
+@Setter
+@Getter
+public class Customer extends BaseEntity {
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Customer.java]: CustomerEntity

File: SQLservise/src/main/java/com/itea/tasks/entities/Developer.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+import javax.persistence.Table;
+import java.util.Set;
+
+
+@Entity
+@Table(name = "developers")
+@NoArgsConstructor
+@Setter
+@Getter
+public class Developer extends BaseEntity {
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Developer.java]: DeveloperEntity

File: SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQuery.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
+package com.itea.tasks.entities;
+
+import javax.persistence.EntityManager;
+import javax.persistence.Persistence;
+import javax.persistence.Query;
+
+public class EntitiesQuery {
+
+    private static EntityManager em = Persistence.createEntityManagerFactory("postgres").createEntityManager();
+    private static Query query;
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQuery.java]: Could you move this variable to each method? you don't need to make it as global

File: SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQueryTest.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
+package com.itea.tasks.entities;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.Assert.assertNotNull;
+
+public class EntitiesQueryTest {
+
+    private static EntitiesQuery entitiesQuery = new EntitiesQuery();
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/EntitiesQueryTest.java]: You can remove static

File: SQLservise/src/main/java/com/itea/tasks/entities/Project.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "projects")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Project.java]: project

File: SQLservise/src/main/java/com/itea/tasks/entities/Project.java [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    @Column(name = "industry")
+    private String industry;
+
+    @Column(name = "cost")
+    private Double cost;
+
+    @ManyToMany(fetch = FetchType.LAZY, mappedBy = "projects")
+    private Set<Developer> developers;
+
+    @ManyToMany(fetch = FetchType.LAZY, mappedBy = "projects")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Project.java]: In @ManyToMany there is a default property for "fetch". So you can remove `fetch = FetchType.LAZY` at all

File: SQLservise/src/main/java/com/itea/tasks/entities/Skill.java [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "skills")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/Skill.java]: skill

File: SQLservise/src/main/java/com/itea/tasks/entities/DeveloperEntity.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+import javax.persistence.Entity;
+import javax.persistence.JoinColumn;
+import javax.persistence.JoinTable;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+
+@Entity
+@Table(name = "developers")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/DeveloperEntity.java]: developer

File: SQLservise/src/main/java/com/itea/tasks/entities/CustomerEntity.java [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.JoinColumn;
+import javax.persistence.JoinTable;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "customers")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/CustomerEntity.java]: customer

File: SQLservise/src/main/java/com/itea/tasks/entities/ProjectEntity.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Setter;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "projects")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/ProjectEntity.java]: project

File: SQLservise/src/main/java/com/itea/tasks/entities/SkillEntity.java [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.Setter;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import java.util.Set;
+
+@Entity
+@Table(name = "skills")
```
Comment (Aleksandr04) [inline SQLservise/src/main/java/com/itea/tasks/entities/SkillEntity.java]: skill

File: SQLservise/src/main/resources/init.sql [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+);
+
+CREATE TABLE skills (
+    id          bigint,
+    industry    varchar(32),
+    skill_level varchar(32),
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE projects(
```
Comment (Aleksandr04) [inline SQLservise/src/main/resources/init.sql]: project

File: SQLservise/src/main/resources/init.sql [hunk 26]
```diff
@@ -0,0 +1,64 @@
+CREATE TABLE developers (
+    id     bigint,
+    name   varchar(32),
+    age    int,
+    gender varchar(32),
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE skills (
```
Comment (Aleksandr04) [inline SQLservise/src/main/resources/init.sql]: skill

File: SQLservise/src/main/resources/init.sql [hunk 27]
```diff
@@ -0,0 +1,64 @@
+CREATE TABLE developers (
```
Comment (Aleksandr04) [inline SQLservise/src/main/resources/init.sql]: developer

File: SQLservise/src/main/resources/init.sql [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+);
+
+CREATE TABLE company (
+    id       bigint,
+    title    varchar(124),
+    industry varchar(32),
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE customers (
```
Comment (Aleksandr04) [inline SQLservise/src/main/resources/init.sql]: customer

**Pull request merged**